chore: Add possibility to validate snippets of any bundle

raffaelecarelle opened the pull request:
Add bundle-specific snippet validation and handling (Storefront and Administration by default)

This update allows snippet validation and retrieval to be scoped to specific bundles via a new `--bundle` option. It introduces methods for processing bundle-specific snippet files and updates the validator interface and implementation to support this feature.

raffaelecarelle committed:
Add bundle-specific snippet validation and handling

This update allows snippet validation and retrieval to be scoped to specific bundles via a new `--bundle` option. It introduces methods for processing bundle-specific snippet files and updates the validator interface and implementation to support this feature.

raffaelecarelle committed:
Refactor snippet handling to simplify methods.

Consolidated snippet file discovery logic by removing specific methods for storefront and administration snippets. Updated relevant tests and adjusted parameter types for better clarity and flexibility. This change improves maintainability and reduces code duplication.

raffaelecarelle committed:
Fix incorrect usage of array_merge in SnippetFileHandler

Corrected the use of array_merge to merge arrays properly. This resolves a potential runtime error when combining storefront and administration snippets.

mitelg commented on src/Core/System/Snippet/SnippetFileHandler.php:
```diff
@@ -29,31 +28,22 @@ public function writeJsonFile(string $path, array $content): void
         file_put_contents($path, $json);
     }
 
-    public function findAdministrationSnippetFiles(): array
```
removing this method is a break.

raffaelecarelle replied:
yes, sorry. reverted

mitelg commented on src/Core/System/Snippet/SnippetValidator.php:
```diff
@@ -48,13 +47,23 @@ public function validate(): array
+    /**
+     * @param array<int, string> $bundles
+     */
+    protected function getAllFiles(array $bundles): SnippetFileCollection
```
adding a required parameter is a break

raffaelecarelle replied:
yes, sorry. I fix it with optional one

mitelg commented on src/Core/System/Snippet/SnippetValidatorInterface.php:
```diff
@@ -7,5 +7,10 @@
+     *
+     * @return array<string, mixed>
+     */
+    public function validate(array $bundles): array;
```
this is also a break

raffaelecarelle replied:
yes, sorry. I fix it with optional one

mitelg replied:
that is still a break :see_no_evil: you cannot simply change the method definition in an interface

raffaelecarelle committed:
fix BC

raffaelecarelle committed:
fix cs

MartinKrzykawski commented on src/Core/System/Snippet/SnippetValidator.php:
```diff
@@ -48,13 +47,23 @@ public function validate(): array
-        $deprecatedFiles = $this->findDeprecatedSnippetFiles();
-        $administrationFiles = $this->snippetFileHandler->findAdministrationSnippetFiles();
-        $storefrontSnippetFiles = $this->snippetFileHandler->findStorefrontSnippetFiles();
+        if ($bundles !== []) {
```
Using `!empty($bundles)` is semantically clearer, IMO.

mitelg replied:
I would like to not use `empty` if possible. it has so many implications :grin: 

I like explicit checks like this more

raffaelecarelle committed:
Refactor snippet validation to support bundle-specific checks

Introduced `BundleSnippetValidatorInterface` to extend snippet validation for specific bundles. Updated `SnippetValidator` to implement the new interface and adjusted dependencies accordingly. This enhances modularity and prepares the system for better bundle-specific validation functionality.

raffaelecarelle committed:
fix cs

MartinKrzykawski commented on src/Core/System/Snippet/SnippetFileHandler.php:
```diff
@@ -29,6 +30,27 @@ public function writeJsonFile(string $path, array $content): void
+    /**
+     * @return array<int, string>
+     */
+    public function findBundleSnippetFiles(BundleInterface $bundle): array
```
Please cover this public function with unit tests too

MartinKrzykawski commented on src/Core/System/Snippet/SnippetFileHandler.php:
```diff
@@ -29,6 +30,27 @@ public function writeJsonFile(string $path, array $content): void
+    }
+
+    /**
+     * @return array<int, string>
```
This function should be deprecated if it is not required anymore.

MartinKrzykawski commented on src/Core/System/Snippet/SnippetFileHandler.php:
```diff
@@ -38,6 +60,9 @@ public function findAdministrationSnippetFiles(): array
     }
 
+    /**
+     * @return array<int, string>
```
This function should be deprecated if it is not required anymore.

MartinKrzykawski commented on src/Core/System/Snippet/SnippetValidator.php:
```diff
@@ -5,26 +5,25 @@
 
 #[Package('discovery')]
-class SnippetValidator implements SnippetValidatorInterface
+class SnippetValidator implements BundleSnippetValidatorInterface
```
This is a break

mitelg replied:
is it? :thinking:

MartinKrzykawski replied:
ups, overlooked that BundleSnippetValidatorInterface extends the SnippetValidatorInterface :D

MartinKrzykawski reviewed (changes requested):
Did you test your changes? I tried running the command with your changes, but it always tells me that both the bundle and the core are valid, even after I deleted snippets and added issues. On trunk, however, it correctly throws validation errors.

mitelg commented on src/Core/System/DependencyInjection/snippet.xml:
```diff
@@ -17,14 +17,22 @@
             <argument type="service" id="Shopware\Core\System\Snippet\Files\SnippetFileCollection" />
             <argument type="service" id="Shopware\Core\System\Snippet\SnippetFileHandler" />
             <argument>%kernel.project_dir%/</argument>
+            <argument type="service" id="kernel"/>
```
should we deprecate this service?

mitelg commented on src/Core/System/Snippet/SnippetValidator.php:
```diff
@@ -48,13 +47,23 @@ public function validate(): array
+    /**
+     * @param array<int, string> $bundles
+     */
+    protected function getAllFiles(array $bundles = []): SnippetFileCollection
```
this is still a break. if someone is extending this class and is overwriting this method, it will break.

mitelg commented on src/Core/System/Snippet/BundleSnippetValidatorInterface.php:
```diff
@@ -0,0 +1,16 @@
+     *
+     * @return array<string, mixed>
+     */
+    public function validate(array $bundles = []): array;
```
I am not sure, if should use the same method name. for a better deprecation path, I think a new method would be better. the other interface should be removed on the long run, right?

raffaelecarelle commented:
Thank you guys, I vert apereciated you help. I give another try. Let me know what you think.

mitelg commented on Makefile:
please remove again :grin:

raffaelecarelle replied:
OMG !

raffaelecarelle replied:
sorry again

mitelg commented on src/Core/System/DependencyInjection/snippet.xml:
```diff
@@ -14,17 +14,26 @@
         </service>
 
         <service id="Shopware\Core\System\Snippet\SnippetValidatorInterface" class="Shopware\Core\System\Snippet\SnippetValidator">
+            <deprecated package="core" version="6.7.0.0">The %service_id% service is deprecated. Use Shopware\Core\System\Snippet\BundleSnippetValidatorInterface instead</deprecated>
```
```suggestion
            <deprecated package="core" version="6.8.0.0">The %service_id% service is deprecated. Use Shopware\Core\System\Snippet\BundleSnippetValidatorInterface instead</deprecated>
```

mitelg commented on src/Core/System/Snippet/BundleSnippetValidator.php:
```diff
@@ -0,0 +1,169 @@
+    /**
+     * @param array<int, string> $bundles
+     */
+    protected function getFileFromBundles(array $bundles = []): SnippetFileCollection
```
does this method have to be protected? I would suggest to make it private, to lower the public API. as we are introducing this completely new

mitelg commented on src/Core/System/Snippet/BundleSnippetValidatorInterface.php:
```diff
@@ -0,0 +1,16 @@
+interface BundleSnippetValidatorInterface
+{
+    /**
+     * @param array<int, string> $bundles
```
```suggestion
     * @param list<string> $bundles
```

mitelg commented on src/Core/System/Snippet/SnippetFileHandler.php:
```diff
@@ -29,6 +30,29 @@ public function writeJsonFile(string $path, array $content): void
+    }
+
+    /**
+     * @deprecated
```
please write the proper deprecation tag. you can find other examples in the code.
something like:

```suggestion
     * @deprecated tag:v6.8.0 - Will be removed. Use `SnippetFileHandler::findBundleSnippetFiles(['administration'])` instead
```

or how ever the replacement would work now :slightly_smiling_face: 

same for the `findStorefrontSnippetFiles`

Additionally you need to trigger a deprecation notice within the method. 

```
        Feature::triggerDeprecationOrThrow(
            'v6.8.0.0',
            Feature::deprecatedMethodMessage(__CLASS__, __METHOD__, 'v6.8.0.0', 'SnippetFileHandler::findBundleSnippetFiles(['administration'])'),
        );
```

Disclaimer: code not tested :wink:

mitelg commented on src/Core/System/Snippet/SnippetValidator.php:
```diff
@@ -6,7 +6,7 @@
 use Shopware\Core\System\Snippet\Files\SnippetFileCollection;
 
-#[Package('discovery')]
+#[Package('framework')]
```
why was that changed?

mitelg commented on src/Core/System/Snippet/SnippetValidator.php:
```diff
@@ -6,7 +6,7 @@
 
-#[Package('discovery')]
+#[Package('framework')]
 class SnippetValidator implements SnippetValidatorInterface
```
this class also needs a deprecation annotation, also the interface file

mitelg commented on tests/unit/Core/System/Snippet/SnippetValidatorTest.php:
```diff
@@ -12,7 +12,7 @@
  * @internal
  */
-#[Package('discovery')]
+#[Package('framework')]
```
why was this changed?

mitelg reviewed (commented):
I had some technical requests. from a domain perspective @MartinKrzykawski needs to say, if there is something missing

raffaelecarelle committed:
Refactor Snippet Validation and Add Deprecations

Refactored snippet validation to separate handling of administration and storefront files, removing bundle dependencies. Deprecated legacy methods and introduced `BundleSnippetValidator` with improved structure and logic. Added unit tests to cover new validation behavior.

raffaelecarelle committed:
Deprecate legacy snippet services and update accessibility.

Marked `SnippetValidator` and related interfaces as deprecated, referencing replacements for version 6.8.0. Updated parameter typing, visibility, and package assignments to align with new conventions and improve clarity. Adjusted tests to reflect these changes.

mitelg commented on src/Core/System/Snippet/Command/ValidateSnippetsCommand.php:
```diff
@@ -27,20 +27,26 @@ class ValidateSnippetsCommand extends Command
     {
-        $missingSnippetsArray = $this->snippetValidator->validate();
+        /** @var list<string> $bundles */
+        $bundles = array_map('trim', array_filter(explode(',', (string) $input->getOption('bundle'))));
```
```suggestion
        $bundles = array_values(array_map('trim', array_filter(explode(',', (string) $input->getOption('bundle')))));
```

then you do not need the var annotation anymore

raffaelecarelle committed:
Fix array index preservation in bundle option processing

Ensure array indexes are reset when mapping and filtering the 'bundle' option input. This prevents unexpected behavior when validating bundles due to non-sequential array keys.

raffaelecarelle committed:
Add DisabledFeatures annotation to snippet validation tests

This change marks specific tests with the DisabledFeatures annotation to ensure compatibility with version v6.8.0.0. It improves test clarity and ensures proper handling of feature toggles.

raffaelecarelle commented:
@mitelg I'm not sure the [Integration tests / 8.2 Storefront mysql:8.0 (pull_request)](https://github.com/shopware/shopware/actions/runs/13588302855/job/37988143691?pr=7072) is related to this PR

MartinKrzykawski reviewed (changes requested):
Hey, thanks for your effort! 

2 things are left:

1. a changelog where you document your change and the deprecations
2. I tested your changes and the validate snippets command is still not working. The output is always that my snippets are valid although i deleted snippets and added invalid json into the storefront, admin and plugin snippets.

mitelg commented on src/Core/System/Snippet/Command/ValidateSnippetsCommand.php:
```diff
@@ -27,20 +27,26 @@ class ValidateSnippetsCommand extends Command
     protected function execute(InputInterface $input, OutputInterface $output): int
     {
-        $missingSnippetsArray = $this->snippetValidator->validate();
+        /** @var list<string> $bundles */
```
```suggestion
```

raffaelecarelle committed:
Remove unnecessary type annotation for bundles variable

The removed type annotation was redundant and added no functional value. This change simplifies the code, making it cleaner and easier to read without altering behavior.

marcelbrode commented:
Hi there, since there's no changes in the last 3 weeks, no further communication and open thread I wanna ask: What's the status here? Can this be reviewed? Is this more of a draft?

raffaelecarelle commented:
> Hi there, since there's no changes in the last 3 weeks, no further communication and open thread I wanna ask: What's the status here? Can this be reviewed? Is this more of a draft? If get

hello, Sorry about that, I put it in draft and then when I've more time try to finish it. thank you

MartinKrzykawski commented:
Hey @raffaelecarelle, thanks for your pull request! Since there hasn’t been any activity for a while, we’re going to close this PR for now. If you find time to complete it in the future, feel free to open a new one — we’d be happy to review it again. 🙂

Best regards!
